hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+
+![Hash Function Example](../assets/hash-tables_hash-tables_hash-function.png)
+
+<!-- Hash function image source: https://en.wikipedia.org/wiki/Hash_function -->
+
+### !callout-info
+
+## Hash Tables Implement Associative Arrays
+
+One reason why hash tables are interesting is because they implement an _abstract data type_ known as an _associative array_. An _abstract data type_ is any data structure that is theoretical, and described not by its concrete implementation, but by its behavior and how it's used. An _associative array_ is a theoretical data structure that describes collections that store key-value pairs.
+
+### !end-callout
+
+## Time Complexity
+
+For all three operations of inserting, finding, and deleting values, hash tables have an O(1) time complexity.
+
+| Action                        | Average Case | Worst Case |
+| ----------------------------- | ------------ | ---------- |
+| Finding a value by key        | O(1)         | O(n)       |
+| Removing a value by key       | O(1)         | O(n)       |
+| Inserting a new key and value | O(1)         | O(n)       |
+
+### !callout-info
+
+## We Describe Hash Table Big O by Average Case
+
+Normally, when we estimate the runtime of a data structure's methods with Big O, we assume the worst case. However, a well-designed hash table will attempt to prevent the worst-case scenario from occurring. Therefore when we work with hash tables we **assume the average-case**. The likelihood of the worst case from occurring depends on how well the **hash function** is implemented.
+
+### !end-callout
+
+## Hash Function
+
+**Hash functions** are functions responsible for mapping a given key into an integer index in the internal array.
+
+### Example Hash Function
+
+Imagine that we have a hash table which holds "name"-"job title" key-value pairs. This hash table needs a hash function to transform the "name" into an index.
+
+An example hash function could use the following algorithm:
+
+- Convert the key to a string
+- Grab the length of this string
+- Get modulo of the length of the string and the length of the internal array
+- Use this modulo as the index
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+### !callout-info
+
+## Why Modulo in This Example?
+
+When arrays get created, they get created with a fixed size in memory. This means that they have a maximum length and _can't_ hold an infinite number of things (unless there is an operation which expands the array. Python's list method `append` manages this for us!). Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!
+
+### !end-callout
+
+## Collisions
+
+There is an enormous number of possible keys and a limited-size array. At some point, a hash function will result in some keys being mapped to the same index. When this occurs, it is called a _collision_.
+
```

Comment:
couldn't resist providing a little detour into the logical framework for why we can definitely say that collisions are inevitable. Love the pigeon hole principle!
```suggestion

### !callout-info

## Collisions Are Predicted by the Pigeon Hole Principle

Any time we try to put `n + 1` items into `n` positions, we will be unable to put each item into a unique position. If we had only `n` items, each item could go into each of the `n` positions perfectly. But as soon as we add one more item, it must now go into the same position as a previous item! This observation is referred to as the _pigeon hole principle_ and can be useful for thinking about the performance of certain algorithms.

### !end-callout

```